Exposing Cruise Control metrics

ppatierno opened the pull request:
### Type of change

- Enhancement / new feature

### Description

This PR is about enabling to expose the Cruise Control metrics through Prometheus (via JMX exporter running as agent) and then showing some of them on a related Grafana dashboard.
The metrics are enabled with the `cruiseControl.metrics` property in the `Kafka` resource where the user can specify the JMX exporter configuration (as we already do for Kafka and ZooKeeper metrics).
The following pictures show the Grafana dashboard and some of the metrics I thought were useful to have at least as an initial starting point (more metrics could be added in the future while using Cruise Control overtime).

![cc_metrics_01](https://user-images.githubusercontent.com/5842311/88822620-62d2d180-d1c4-11ea-8000-f515e549acb1.png)
![cc_metrics_02](https://user-images.githubusercontent.com/5842311/88822628-65352b80-d1c4-11ea-8497-fd84c4f9ef5b.png)
![cc_metrics_03](https://user-images.githubusercontent.com/5842311/88822635-66feef00-d1c4-11ea-9093-19b4133e832e.png)
![cc_metrics_04](https://user-images.githubusercontent.com/5842311/88822645-69f9df80-d1c4-11ea-8b4a-ed98058310fc.png)

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [x] Supply screenshots for visual changes, such as Grafana dashboards

tomncooper commented:
LGTM. The only potential issue is that things like the Goal Violation rate are based on the anomaly detector which we haven't mentioned in docs AFAIK? Might be worth us adding "support" for the anomaly detector and adding configuarable anomaly detection goals etc in another PR.

scholzj commented on examples/metrics/grafana-dashboards/strimzi-cruise-control.json:
```diff
@@ -0,0 +1,2062 @@
+        "#d44a3a"
+      ],
+      "datasource": "${DS_PROMETHEUS}",
+      "description": "Balancedness Score",
```
Can we get some better descriptions? For example I have no idea what does _Balancedness Score_, which value is good, which bad etc. ... I guess 100 is bad here since it is orange in the screenshot?

ppatierno replied:
Yes, I reverted the logic in the dashboard, now higher values are green while lower are orange and then red. I used 50 (for red) and 80 (orange) as thresholds.

scholzj replied:
So is 100 a maximum? TBH, I think the Cruise Control is relative niche ... so the descriptions are IMHO quite important since many people will be looking at it as I'm ... without having a clue.

BTW: Should we have a sample alert for this as well?

ppatierno replied:
Yes 100 is the max.
Regarding the alerts I would add them in a different PR as already did for the bridge metrics.

tomncooper replied:
I think it is useful to have this metric as a one-stop-shop for quickly checking the state of the cluster. To be clear this is the current balancedness score of the cluster as calculated by the anomaly detector (using `KafkaCruiseControlUtils.balancednessCostByGoal` CC method) and isupdated every 5 mins by default (can be changed by setting `anomaly.detection.interval.ms`). I wonder if we need some docs to describe how this is calculated and what it means?

scholzj replied:
So can we add this stuff to the description?

scholzj commented on examples/metrics/grafana-dashboards/strimzi-cruise-control.json:
```diff
@@ -0,0 +1,2062 @@
+        "#bf1b00"
+      ],
+      "datasource": "${DS_PROMETHEUS}",
+      "description": "Number of valid windows",
```
What does valid window mean?

ppatierno replied:
It should be a time window where there are enough samples for computing a proposal, but I am sure @tomncooper can say more.

tomncooper replied:
Each window has a number of samples after which it is considered "valid". The user can also configure how many windows CC should hold onto. Details [here](https://github.com/linkedin/cruise-control/wiki/Overview#metric-sample-aggregator)

scholzj replied:
So can we add it to the description?

scholzj commented on examples/metrics/grafana-dashboards/strimzi-cruise-control.json:
```diff
@@ -0,0 +1,2062 @@
+        "#299c46"
+      ],
+      "datasource": "${DS_PROMETHEUS}",
+      "description": "The number of snapshot windows that is monitored",
```
Should it be are monitored? (just guessing)

```suggestion
      "description": "The number of snapshot windows that is monitored",
```

ppatierno replied:
You are suggesting the same :-) ... but I got what you want ;-)

scholzj commented on examples/metrics/grafana-dashboards/strimzi-cruise-control.json:
```diff
@@ -0,0 +1,2062 @@
+      "dashLength": 10,
+      "dashes": false,
+      "datasource": "${DS_PROMETHEUS}",
+      "description": "Disk Failure Rate",
```
What does this mean? That the disk died?

ppatierno replied:
So as far as I understood from the DiskFailureDetector code, it periodically tries to get info on the log dirs on the brokers (using the Admin Client API `describeLogDirs`) and the callback is able to provide info about errors which are potential disk failures.

scholzj replied:
I think this is a bit scary then ... what about calling it _Failed disk queries_ or something like that?

ppatierno replied:
Maybe "disk read failure rate"?

scholzj replied:
TBH, I think _Failed disk queries_ would be better because IMHO it makes it more clear that this is just about Cruise Control issue when querying the disk. _Disk read failure rate_ can still be easily confused with a physical disk issue. But I can live with it if needed.

ppatierno replied:
I see your point you want to highlight that it's something more related to CC than the disk itself even if when CC has a "query" failure, it could mean a real disk failure. @tomncooper any opinion on this?

tomncooper replied:
So the disk failures are detected using the code [here](https://github.com/linkedin/cruise-control/blob/a7ff697af10f20c881239dfe41283460965edbaf/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/DiskFailureDetector.java#L89). It seems to use the Kafka Admin client to ask for broker status and then checks for reported errors in each of the log directories.

tomncooper replied:
So this looks to be a replication of the Kafka side metrics?

ppatierno replied:
That was the code I was referring to (without linking the code actually). But what's a good name/description for you?

tomncooper replied:
Sorry @ppatierno I didn't read your comment at the top and basically repeated exactly what you said. Not very helpful. Anyway this metric is the number of disc failures in the last detection interval (which will be the same as the AD interval unless the user specifies a different interval for the disk failure detection), so it is a "rate" and it is a failure according to Kafka. So "Disk Failure Rate"/"Reported Disk Failure Rate" seems pretty accurate to me.

scholzj commented on examples/metrics/grafana-dashboards/strimzi-cruise-control.json:
```diff
@@ -0,0 +1,2062 @@
+      "dashLength": 10,
+      "dashes": false,
+      "datasource": "${DS_PROMETHEUS}",
+      "description": "Rebalance Request Rate",
```
Is this number of executed partition movemenets per second or something like that?

ppatierno replied:
It's the REST API call to the `/rebalance` endpoint.

ppatierno commented:
> You should probably also update the metrics docs etc. (e.g. https://strimzi.io/docs/operators/latest/full/deploying.html#ref-metrics-config-files-str etc.)

I will do when the review will pass in terms of metrics if we should have a dedicated paragraph as we have for the Bridge.

> Do we need to add CC to PodMonitor? Or it is already there?

The current `kafka-metrics` PodMonitor is already able to scrape CC because it matches the following:

```yaml
matchExpressions:
      - key: "strimzi.io/kind"
        operator: In
        values: ["Kafka", "KafkaConnect"]
``` 

I should rename the PodMonitor as `kafka-and-cruise-control-metrics`.

ppatierno commented:
> LGTM. The only potential issue is that things like the Goal Violation rate are based on the anomaly detector which we haven't mentioned in docs AFAIK? Might be worth us adding "support" for the anomaly detector and adding configurable anomaly detection goals etc in another PR.

If you think that it's better, I can even delete this metric for now and we will re-add it later.

tomncooper commented on examples/metrics/grafana-dashboards/strimzi-cruise-control.json:
```diff
@@ -0,0 +1,2062 @@
+      "description": "Number of valid windows",
+      "format": "none",
+      "gauge": {
+        "maxValue": 100,
```
AFAIK this value is not always going to 100. The user can define how many windows that CC should hold on to.

ppatierno replied:
It's just a default provided by Grafana, it doesn't hurt I guess.

samuel-hawker commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CruiseControl.java:
```diff
@@ -160,6 +164,7 @@ protected CruiseControl(HasMetadata resource) {
         this.mountPath = "/var/lib/kafka";
         this.logAndMetricsConfigVolumeName = LOG_AND_METRICS_CONFIG_VOLUME_NAME;
         this.logAndMetricsConfigMountPath = LOG_AND_METRICS_CONFIG_VOLUME_MOUNT;
+        this.isMetricsEnabled = DEFAULT_CRUISE_CONTROL_METRICS_ENABLED;
```
I'm not sure if this boolean needs to exist if we are always equating it to 
```
 if (metrics != null)
```

Could always just gate the logic on that? (i.e. line 600)

ppatierno replied:
I agree but this field comes from history, I would avoid to change it now because removing it means refactoring other classes still using the field and they are not related to this PR. We should do a refactoring for that in a different PR.

tomncooper commented:
> If you think that it's better, I can even delete this metric for now and we will re-add it later.

I think we need to go over the AnomalyDetector configs and add some docs around configuring the goals used for the anomaly detector (the violation of which lead to this metric). I'll leave it up to you if we want to do that before we enable this metric.

ppatierno commented:
@tomncooper so currently we don't allow to specify the anomaly detector goals in the CC configuration but we are using the default ones.
I think that it makes sense to leave the related metrics here (goal violation, balancedness score, and disk failure) because they are interesting ones and leaving for a different PR the enabling of the anomaly detector goals configuration.
Meanwhile, the user can get the related metrics values for the default anomaly detector goals.

ppatierno commented:
@tomncooper @scholzj @samuel-hawker @kyguy I fixed comments (I hope) and added documentation (more ore less the same way we did for the bridge). Can you have another pass, please?

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+
+The example Grafana dashboard for the Cruise Control provides:
+
+* Information about rebalance proposals computation, goals violation, cluster balancedness and more
```
This should be "optimization proposals" rather than "rebalance" to be consitent with the rest of the docs.

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+Cruise Control:: Shows metrics for:
++
+* The number of snapshot windows that are monitored
+* Number of time windows considered to be valid because of enough samples for computing a proposal.
```
The other points don't end with a full stop so, for consitency, you should remove the one at the end of this line.

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
++
+* The number of snapshot windows that are monitored
+* Number of time windows considered to be valid because of enough samples for computing a proposal.
+* Number of ongoing executions running for proposals or rebalance
```
"rebalances"

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* Number of time windows considered to be valid because of enough samples for computing a proposal.
+* Number of ongoing executions running for proposals or rebalance
+* The current balancedness score of the Kafka cluster as calculated by the anomaly detector (every 5 minutes by default)
+* The percentafe of monitored partitions
```
"percentage"

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* Number of ongoing executions running for proposals or rebalance
+* The current balancedness score of the Kafka cluster as calculated by the anomaly detector (every 5 minutes by default)
+* The percentafe of monitored partitions
+* How often a goal violation happens
```
Is it a rate? Probably better to say "the number of goal violations reported by the anomaly detector (every 5 mins by default)".

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* The percentafe of monitored partitions
+* How often a goal violation happens
+* How often a disk read failure happens on the brokers
+* Failures rate about fetching samples
```
"Rate of metric sample fetch failures"

tomncooper commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* How often a goal violation happens
+* How often a disk read failure happens on the brokers
+* Failures rate about fetching samples
+* Time needed for computing a rebalance proposal
```
optimization proposal

ppatierno commented:
/azp run system-tests

laidan6000 commented on documentation/modules/metrics/proc_metrics-grafana-dashboard.adoc:
```diff
@@ -15,6 +15,7 @@ Example dashboards are provided in the `examples/metrics` directory as JSON file
+* `strimzi-cruise-control.json`
 
 The example dashboards are a good starting point for monitoring key metrics, but they do not represent all available metrics.
 You can modify the example dashboards or add other metrics, depending on your infrastructure.
```
At the end of this module there are bullet lists showing the metrics on each dashboard.

I suggest adding a new metrics list for Cruise Control.

laidan6000 replied:
Actually, that information is in the new assembly. I'm not sure you need to repeat it again.

ppatierno replied:
Yeah the list is already in the new assembly as for the bridge, I don't think we should repeat it again.

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+
+If you are already using Prometheus and Grafana for monitoring of built-in Kafka metrics, you can configure Prometheus to also scrape the Cruise Control Prometheus endpoint.
+
+The example Grafana dashboard for the Cruise Control provides:
```
```suggestion
The example Grafana dashboard for Cruise Control provides:
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+
+The example Grafana dashboard for the Cruise Control provides:
+
+* Information about optimization proposals computation, goals violation, cluster balancedness and more
```
```suggestion
* Information about optimization proposals computation, goals violation, cluster balancedness, and more
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+
+== Configuring Cruise Control
+
+You can enable the Cruise Control metrics in the `Kafka` resource using the `cruiseControl.metrics` property that contains the JMX exporter configuration about metrics to expose.
```
```suggestion
You can enable the Cruise Control metrics in the `Kafka` resource using the `cruiseControl.metrics` property that contains the JMX exporter configuration about the metrics to expose.
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+You can enable the Cruise Control metrics in the `Kafka` resource using the `cruiseControl.metrics` property that contains the JMX exporter configuration about metrics to expose.
+
+For example:
++
```
You don't need a plus here unless I'm mistaken!

ppatierno replied:
I copied it from the assembly_metrics-kafka-bridge.adoc, if you think it's not needed I am going to remove it anyway.

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+
+Cruise Control:: Shows metrics for:
++
+* The number of snapshot windows that are monitored
```
```suggestion
* Number of snapshot windows that are monitored by Cruise Control
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+Cruise Control:: Shows metrics for:
++
+* The number of snapshot windows that are monitored
+* Number of time windows considered to be valid because of enough samples for computing a proposal
```
```suggestion
* Number of time windows considered valid because they contain enough samples to compute an optimization proposal
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* The number of snapshot windows that are monitored
+* Number of time windows considered to be valid because of enough samples for computing a proposal
+* Number of ongoing executions running for proposals or rebalances
+* The current balancedness score of the Kafka cluster as calculated by the anomaly detector (every 5 minutes by default)
```
```suggestion
* Current balancedness score of the Kafka cluster as calculated by the anomaly detector component of Cruise Control (every 5 minutes by default)
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* Number of time windows considered to be valid because of enough samples for computing a proposal
+* Number of ongoing executions running for proposals or rebalances
+* The current balancedness score of the Kafka cluster as calculated by the anomaly detector (every 5 minutes by default)
+* The percentage of monitored partitions
```
```suggestion
* Percentage of monitored partitions
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* Number of ongoing executions running for proposals or rebalances
+* The current balancedness score of the Kafka cluster as calculated by the anomaly detector (every 5 minutes by default)
+* The percentage of monitored partitions
+* The number of goal violations reported by the anomaly detector (every 5 minutes by default)
```
```suggestion
* Number of goal violations reported by the anomaly detector (every 5 minutes by default)
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* The number of goal violations reported by the anomaly detector (every 5 minutes by default)
+* How often a disk read failure happens on the brokers
+* Rate of metric sample fetch failures
+* Time needed for computing an optimization proposal
```
```suggestion
* Time needed to compute an optimization proposal
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* How often a disk read failure happens on the brokers
+* Rate of metric sample fetch failures
+* Time needed for computing an optimization proposal
+* Time for creating a cluster model
```
```suggestion
* Time needed to create the cluster model
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* Rate of metric sample fetch failures
+* Time needed for computing an optimization proposal
+* Time for creating a cluster model
+* How often a proposal request or an actual rebalance request is made through the REST API
```
```suggestion
* How often a proposal request or an actual rebalance request is made through the Cruise Control REST API
```

laidan6000 commented on documentation/assemblies/metrics/assembly_metrics-cruise-control.adoc:
```diff
@@ -0,0 +1,69 @@
+* Time needed for computing an optimization proposal
+* Time for creating a cluster model
+* How often a proposal request or an actual rebalance request is made through the REST API
+* How often the overall cluster state and the user tasks state are requested through the REST API
```
```suggestion
* How often the overall cluster state and the user tasks state are requested through the Cruise Control REST API
```

ppatierno commented:
@samuel-hawker are you ok with changes? do you need another review pass?
